### DISCUSSION TITLE
Advice for beamforming and setting up mic array

### DISCUSSION THREAD

**Discussion by IsamarZhu** [acoular/acoular#350] [Q&A] [answered]
Hello,

I'm currently trying to perform beamforming on a longer snippet of audio by breaking it up into chunks to visualize how the sound source moves. My audio contains beeps of 500Hz, 1000Hz, 2000Hz, 5000Hz, and 10000Hz, all 10 seconds each and at different locations within the mic array. My mic array consists of 8 microphones (8 channels), where the mics are at corners of a 12 ft x 12 ft square, and each pair of mics are spaced 0.144 m apart.

Currently, I set up my mics so that the center of the square is at (0, 0), and the x-y plane is parallel with the ground.

If anyone could help debug my setup, that would be much appreciated! My code and sample outputs are below. I'm not too sure where the ripples/dots in the output are coming from, and I'm not able to capture the location of the sound accurately.

- BeamformerBase vs BeamformerCapon
- The z-axis in `rg = acoular.RectGrid(x_min=-3, x_max=3, y_min=-3, y_max=3, z=-0.3, increment=0.01)`, since the speaker is 14 in. shorter than the mics
- Setup of the mic_array (are the mics spaced too far apart to capture the frequency of the beeps?)
-  The block size

Any help would be appreciated, thanks!!

https://github.com/IsamarZhu/AcoularBuzzCam/blob/main/mic_array/video.py
https://github.com/IsamarZhu/AcoularBuzzCam/blob/main/mic_array/mic_array.xml

**Comment by artpelling** [comment, marked as answer]
Hi @IsamarZhu,

> If anyone could help debug my setup, that would be much appreciated! 

Absolutely!

I've briefly taken a look at your code and I do not immediately see anything issues with it.

The problem you report regarding the accuracy of localization is probably due to your array geometry which is maybe a little unconventional. Most commonly, planar microphone arrays for 2d sound localization would consist of a larger number of microphones that are distributed across the entire array plane (like a grid or maybe a logarithmic spiral). 

> -  Setup of the mic_array (are the mics spaced too far apart to capture the frequency of the beeps?)

While "far" is always depends on the frequency and spatial range of interest, yes, I would say they are quite far apart. Moreover, it might be good to think about the distribution of inter-microphone distances in your array. Your array plane is quite "large" (3.6m x 3.6m) with the microphones in the corners. Thus, you have quite "large" distances between the microphones together with "smaller" ones of 14cm.

> I'm not too sure where the ripples/dots in the output are coming from, and I'm not able to capture the location of the sound accurately.

What you are seeing in the output is probably your array's point spread function (PSF) which as indicated earlier is probably not ideal for your spatial area and frequency range of interest. I am sure that there are other constraints limiting the placement of microphones as usual but you could consider altering your array geometry to improve your array's PSF. 

Here is a small code snippet that you could use to visualize your PSF written by @adku1173:

``` python
import acoular as ac
import matplotlib.pyplot as plt
from pathlib import Path

mg = ac.MicGeom(from_file=Path(__file__).parent / 'mgeom.xml')


steer = ac.SteeringVector( 
    grid=ac.RectGrid( x_min=-3.0, x_max=3., y_min=-3.0, y_max=3.0, z=-0.3, increment=0.05), 
    mics=mg )

psf = ac.PointSpreadFunction( steer=steer, calcmode='single', freq=2000)

plt.figure(1, (15, 5))  
for i, increment in enumerate([0.05, 0.1, 0.2]):
    steer.grid.increment = increment
    psf.grid_indices=[steer.grid.size//2]
    psff_center =  ac.L_p(psf.psf[:, 0]).reshape(steer.grid.shape)
    plt.subplot(1, 3, i+1)
    plt.title(f'increment={increment}')
    plt.imshow(psff_center.T, extent=steer.grid.extend(), origin='lower', vmax=psff_center.max(), vmin=psff_center.max()-40)
    plt.colorbar()
plt.show()
```

Alternatively, you could take a look at the `MicGeomExample` app of [spectacouar](https://github.com/acoular/spectacoular/) with which you can visualize your PSF interactively and move microphones around. Here is a [short demo](https://acoular.github.io/spectacoular/_images/micgeomexample.mp4).

Let me know if that helped you and/or if you have any other questions :)

Some other comments:

> since the speaker is 14 in. shorter than the mics

This sentence is a bit unclear to me. Does that mean your sound source is located in the array plane?

> rg = acoular.RectGrid(x_min=-3, x_max=3, y_min=-3, y_max=3, z=-0.3, increment=0.01)

While not really changing the results, choosing a coarser increment, maybe `0.1` or `0.2` might already improve your output visually and would "simplfy" your PSF. Further, the dynamic range you chose was very small.

Lastly, I've removed the google drive links from your original comment for security reasons. It would be good to simply share code (as you've done) or embed any images directly.
